src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java

```diff
@@ -0,0 +1,205 @@
+import com.microsoft.aad.msal4j.ClientCredentialFactory;
+import com.microsoft.aad.msal4j.ClientCredentialParameters;
+import com.microsoft.aad.msal4j.ConfidentialClientApplication;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import com.microsoft.aad.msal4j.IClientCredential;
+import com.microsoft.aad.msal4j.SilentParameters;
+import java.net.MalformedURLException;
+import java.text.MessageFormat;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.HashSet;
+import java.util.concurrent.CompletableFuture;
+import reactor.core.publisher.Mono;
+
+
+/**
+ * An AAD credential that acquires a token with a client secret for an AAD application.
+ */
+@Immutable
+class KeyVaultTokenCredential implements TokenCredential {
+    private final ClientLogger logger = new ClientLogger(KeyVaultTokenCredential.class);
+    private final String clientId;
+    private final String clientSecret;
+    private final SQLServerKeyVaultAuthenticationCallback authenticationCallback;
+    private String authorization;
+    private ConfidentialClientApplication confidentialClientApplication;
+    private String resource;
+    private String scope;
+
+    /**
+     * Creates a KeyVaultCredential with the given identity client options.
+     *
+     * @param clientId
+     *        the client ID of the application
+     * @param clientSecret
+     *        the secret value of the AAD application
+     * @throws SQLServerException
+     */
+    KeyVaultTokenCredential(String clientId, String clientSecret) throws SQLServerException {
+        if (null == clientId || clientId.isEmpty()) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Client ID"};
+            throw new SQLServerException(form.format(msgArgs1), null);
+        }
+
+        if (null == clientSecret || clientSecret.isEmpty()) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Client Secret"};
+            throw new SQLServerException(form.format(msgArgs1), null);
+        }
+
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.authenticationCallback = null;
+    }
+
+    /**
+     * Creates a KeyVaultCredential with the given identity client options.
+     *
+     * @param authenticationCallback The authentication callback that gets invoked when an access token is requested.
+     */
+    KeyVaultTokenCredential(SQLServerKeyVaultAuthenticationCallback authenticationCallback) {
+        this.authenticationCallback = authenticationCallback;
+        this.clientId = null;
+        this.clientSecret = null;
+    }
+
+    @Override
+    public Mono<AccessToken> getToken(TokenRequestContext request) {
+        if (null != authenticationCallback) {
+            // If the callback is not null, invoke the callback to get the token. This gets invoked each time
```

Comment:
Please change this to multi-line comment.

Reply:
Not resolved.

Reply:
sorry you're right, this change isn't merged to this PR yet..